user/tripwire/man/siggen.8
.\" $Id: siggen.8,v 1.9 1994/07/25 15:46:22 gkim Exp $
.\"
.TH SIGFETCH 8 "October 14, 1992"
.SH NAME
siggen \- signature gathering routine for Tripwire
.SH SYNOPSIS
.B siggen
siggen [-0123456789aqv] [ 
.I file 
\&... ]
.SH DESCRIPTION
.LP
\fIsiggen\fP is a utility that fetches signatures used by Tripwire.
.SH OPTIONS
.LP
When run without any arguments, \fIsiggen\fP collects all the signatures
on the specified files in a verbose manner.
.LP
When no arguments are given, \fIsiggen\fP reads from standard input.
.TP 
.B \-h
prints the signature in hexadecimal instead of base-64.
.TP 
.B \-q
Quiet mode.  Prints out all signatures in one line with no extraneous
information.
.TP 
.B \-a
Generates all signatures.  (default)
.TP
.B \-v
Verbose mode.  (default)
.TP
.B \-0
signature 0 - generates null signature
.TP
.B \-1
signature 1 - generates MD5, the RSA Data Security, Inc. Message Digesting Algorithm.
.TP
.B \-2
signature 2 - generates Snefru, the Xerox Secure Hash Function.
.TP
.B \-3
signature 3 - generates CRC-32, POSIX 1003.2 compliant 32-bit Cyclic Redundancy Check.
.TP
.B \-4
signature 4 - generates CRC-16, the standard (non-CCITT) 16-bit Cyclic Redundancy Check.
.TP
.B \-5
signature 5 - generates MD4, the RSA Data Security, Inc. Message Digesting Algorithm.
.TP
.B \-6
signature 6 - generates MD2, the RSA Data Security, Inc. Message Digesting Algorithm.
.TP
.B \-7
signature 7 - generates SHA, the NIST Secure Hash Algorithm (NIST FIPS 180)
.TP
.B \-8
signature 8 - generates Haval, a 128-bit signature code
.TP
.B \-9
signature 9 \- generates null signature (reserved for future expansion)
.SH AUTHORS
.nf
Gene Kim
Purdue University
gkim@cs.purdue.edu
.fi




freeswan/CHANGES
New and notable in 1.97:

Sanity check added for at least one of AH or ESP more for checking if
"make *config" has even been run.

Fixed 2.2 local IKE fragmentation blackhole.  Still won't work if
iptraf or another pcap app is running.

MCR added ikeping(8).

Applied DHR's tunnel patch to streamline IKE/specialSA processing.

Added local and remote IKE bypass tests to the UML test suite.

UML KLIPS %trap* tests now check for upbound pf_key messages.

An ICMP ICMP_DEST_UNREACH, ICMP_PKT_FILTERED is now sent upon %reject.

Added a UML input file with first and last packets being outside the 
covered range to "sandwich" the packets under test.

The default _updown script has been cleaned up slightly, reducing
internal duplication that arose accidentally.

When doing --add on an RSA-signature conn, ipsec_auto now tells Pluto
about the keys before, not after, telling it about the conn.  The main
consequence of this is that if something is wrong with one of the keys,
the conn itself does *not* get loaded.  This should reduce confusion.

The rp_filter checks during KLIPS startup have been fixed to be a bit less
paranoid; rp_filter is thought to cause problems only on the physical
interfaces, not on the ipsecN interfaces. 

The configuration process has been fixed to be a bit more paranoid.  Some
users have had mysterious compile failures with complaints involving names
like MD5_CTX; they seem to have resulted from kernel configuration changes
somehow getting into .config but not into include/linux/autoconf.h (which
is what actually controls most of the kernel compile).  The post-config
checking now verifies not only that .config has the right stuff in it, but
that autoconf.h does too.  (We still don't see *how* this discrepancy can
arise, but this will at least catch it when it does.)

The details of linking the KLIPS source into the kernel are now under
control of a configuration variable in Makefile.inc, to make it easier
for distribution builders to change.

The top-level Makefile now includes cleverness which assigns a date-based
version code if a build is done from a raw CVS checkout.

The library now includes a new function, ttodatav(), like ttodata() but
in some cases it gives more verbose error reports.  Also, there are now
a couple of functions which distill an RSA public key to a "key ID" for
use in messages, a crypto-quality pseudo-random-number generator, and a
function that converts a single address into a (singleton) subnet.

Barf's secret/key-censoring utilities have been tidied up so that what
they emit matches the key-ID library functions. 

The ipsec command now supplies a couple more environment variables to
things it calls, and this has reduced the number of things which need to
be edited at build time to know about configuration variables. 

The configuration reader now allows parameter names to begin with "_";
these are strictly reserved for internal development purposes and will
not in general be documented.


As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.96:

Compatibility with the 2.0.xx kernels (under Red Hat 5.2) is broken.
This was accidental, and we have some hopes of fixing it, although it
won't last forever even if we do.  See README.

Our "IKE Implementation Issues" IETF draft has been added to our docs.

Newhostkey now has a mandatory --output "option" specifying a filename
(although the magic filename "-" means "standard output"), and sets
restrictive permissions on newly-created files to try to keep private keys
truly private. 

Pluto now gracefully handles a couple of odd error cases (missing private
keys, network errors at just the wrong time in negotiations) which used to
cause it to die, provoking an automatic restart of the IPsec subsystem. 

Makefile.inc now has more detailed control over where manpages go, to ease
integration into distributions.  The name of the destination directory for
RPMs is also now centralized there.

Rsasigkey now checks for certain obscure internal failures, which can
happen if --random is used to supply a source of random bits that aren't
really very random, and complains accordingly.

Barf's eroute table, formerly sorted by source, is now sorted by destination.
This is probably a bit clearer for hosts with lots of eroutes.

Use of the %reject target in a shunt eroute now, as long planned, actually
causes an ICMP rejection message (Destination Unreachable:  Communication
Administratively Prohibited). 

KLIPS now declares its module license to be "Dual BSD/GPL".  (It is 
dual because the libdes and zlib code is BSD licensed, while KLIPS itself 
is GPL).  Whether this is exactly right is not yet clear, but at least it
stops the latest module loaders from whining.

The UML test stuff continues to grow and add new test cases.

As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.95:

The last remnants of the "%hold" bug, which broke 1.93 and 1.94, have
(we think) been dealt with.

Minor bugs in the RPM-building stuff have been fixed, and it has been
updated to deal with Red Hat's abrupt removal of the --buildroot option
from rpm.

Pluto now supports the ability to delete an instance of a conn, rather
than just the whole conn (for cases, notably OE, where a single connection
description can be involved in multiple connections).  There is no support
for this in ipsec_auto yet.

Pluto's routine (non-debug) logging is now somewhat terser, and some of
the more boring messages are produced only if --debug-lifecycle is turned
on (this is a temporary hack, it's not really debugging output).

Pluto's innards have been reorganized somewhat, in preparation for making
DNS lookups asynchronous.  There should be no user-visible effects.

Pluto now ignores IPv6 link-local interface addresses; we think that they
are never relevant, and trying to bind to them causes difficulties.  This

user/tripwire/src/siggen.c
#ifndef lint
static char rcsid[] = "$Id: siggen.c,v 1.14 1994/07/25 15:24:12 gkim Exp $";
#endif

/*
 * siggen.c
 *
 *	generate signatures for a given file.
 *
 * Gene Kim
 * Purdue University
 * October 14, 1992
 */

#include "../include/config.h"
#include <stdio.h>
#include <fcntl.h>
#ifdef STDLIBH
# include <stdlib.h>
#endif
#include <sys/param.h>
#include <sys/types.h>
#include <sys/stat.h>
#ifndef XENIX
# include <sys/time.h>
#else
# include <time.h>
#endif 	/* XENIX */
#ifdef DIRENT
# include <dirent.h>
#else
# ifndef XENIX
#  include <sys/dir.h>
# else		/* XENIX */
#  include <sys/ndir.h>
# endif		/* XENIX */
#endif	/* DIRENT */
#if (defined(SYSV) && (SYSV < 3))
# include <limits.h>
#endif	/* SVR2 */
#ifdef STRINGH
#include <string.h>
#else
#include <strings.h>
#endif
#include "../include/list.h"
#include "../include/tripwire.h"

#ifndef L_tmpnam
# define L_tmpnam (unsigned int) MAXPATHLEN
#endif

extern int optind;
int debuglevel = 0;
char *mktemp();

int (*pf_signatures [NUM_SIGS]) () = {
					SIG0FUNC,
					SIG1FUNC,
					SIG2FUNC,
					SIG3FUNC,
					SIG4FUNC,
					SIG5FUNC,
					SIG6FUNC,
					SIG7FUNC,
					SIG8FUNC,
					SIG9FUNC
				      };

int printhex = 0;
int sigallget = 1;
int sigvector[NUM_SIGS] = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
char *signames[NUM_SIGS] = {
					SIG0NAME,
					SIG1NAME,
					SIG2NAME,
					SIG3NAME,
					SIG4NAME,
					SIG5NAME,
					SIG6NAME,
					SIG7NAME,
					SIG8NAME,
					SIG9NAME,
			   };
int verbosity = 0;
int quietmode = 0;
char *tmpfilename = NULL;
int readstdin = 0;


char *progname;

void
usage()
{
    int i;

    fprintf(stderr, "siggen: usage: [-0123456789qv] [-h] [ file ... ]\n");
    fprintf(stderr, "	(-h is to print signatures in hexadecimal.  default is base-64.\n");
    for (i = 0; i < sizeof(signames)/sizeof(char *); i++) {
	fprintf(stderr, "\tsig %d: %s\n", i, signames[i]);
    }

    exit(1);
}

int
main(argc, argv)
    int argc;
    char *argv[];
{
    int i, c;
    int fd;
    int errors = 0;

    progname = argv[0];

    optind = 1;
    while ((c = getopt(argc, argv, "0123456789aqvh")) != -1) {
	switch(c) {
	case '0':		sigallget = 0; sigvector[0] = 1; break;
	case '1':		sigallget = 0; sigvector[1] = 1; break;
	case '2':		sigallget = 0; sigvector[2] = 1; break;
	case '3':		sigallget = 0; sigvector[3] = 1; break;
	case '4':		sigallget = 0; sigvector[4] = 1; break;
	case '5':		sigallget = 0; sigvector[5] = 1; break;
	case '6':		sigallget = 0; sigvector[6] = 1; break;
	case '7':		sigallget = 0; sigvector[7] = 1; break;
	case '8':		sigallget = 0; sigvector[8] = 1; break;
	case '9':		sigallget = 0; sigvector[9] = 1; break;
	case 'a':		sigallget = 1; break;
	case 'v':		verbosity = 1; break;
	case 'q':		quietmode = 1; break;
	case 'h':		printhex = 1; break;
	case '?':
	default:
	    usage();
	    exit(1);
	}
    }

    argc -= optind;
    argv += optind;

    if (argc == 0)
	readstdin = 1;
    for (i = 0; i < argc; i++) {
	if (strcmp(argv[i], "-") == 0) {
	    readstdin = 1;
	    continue;
	}
	else if ((fd = open(argv[i], O_RDONLY, 0)) < 0) {
	    warn_with_err("siggen: skipping '%s'", argv[i]);
	    errors++;
	    continue;
	}
	if (argc > 1)
	    printf("*** %s ***\n", argv[i]);
	if (siggen(fd) < 0)
	    errors++;

	if (fd)
	    close(fd);
    }

    if (readstdin) {
	FILE *fpout;
	/* generate temporary file name */
	if ((tmpfilename = (char *) malloc(L_tmpnam + MAXPATHLEN)) == NULL) {
	    perror("main: malloc()");
	    exit(1);
	};
	(void) strcpy(tmpfilename, "/tmp/twzXXXXXX");

	if ((char *) mktemp(tmpfilename) == NULL) {
	    perror("siggen: mktemp()");
	    exit(1);
	}

	/*  output */
	if (!(fpout = fopen(tmpfilename, "w"))) {
	    char err[1024];
	    sprintf(err, "main: fopen(%s)", tmpfilename);
	    perror(err);
	    exit(1);
	}
	/*  copy */
	while ((c = getc(stdin)) != EOF)
	    putc(c, fpout);
	fclose(fpout);
	if ((fd = open(tmpfilename, O_RDONLY)) < 0) {
	    perror("siggen: open");
	    exit(1);
	}
	if (siggen(fd) < 0)
	    errors++;

	if (fd)
	    close(fd);
	unlink(tmpfilename);
    }

    if (errors) 
	exit(1);


    exit(0);
}

int
siggen(fd)
    int fd;
{
    char 	sigs[NUM_SIGS][SIG_MAX_LEN];
    int 	i;

    /* collect signatures */
    for (i = 0; i < NUM_SIGS; i++) {
	char *pc = sigs[i];

	if (sigallget || sigvector[i]) {

	    if ((*pf_signatures[i])(fd, pc, SIG_MAX_LEN) < 0) {
	        return -1;
	    }
	    if (!quietmode)
		printf("sig%d: %-9s: %s\n", i, signames[i], sigs[i]);
	    else
		printf("%s ", sigs[i]);

	}
    }

    if (quietmode)
        printf("\n");

    return 0;
}
